client/rest-high-level/src/main/java/org/elasticsearch/client/ml/datafeed/DatafeedConfig.java

```diff
@@ -228,31 +236,35 @@ public boolean equals(Object other) {
             && Objects.equals(this.queryDelay, that.queryDelay)
             && Objects.equals(this.indices, that.indices)
             && Objects.equals(this.types, that.types)
-            && Objects.equals(this.query, that.query)
+            && Objects.equals(asMap(this.query), asMap(that.query))
             && Objects.equals(this.scrollSize, that.scrollSize)
-            && Objects.equals(this.aggregations, that.aggregations)
+            && Objects.equals(asMap(this.aggregations), asMap(that.aggregations))
```

Comment:
Using `asMap()` in `equals()` and `hashCode()` will make them heavyweight operations.  Maybe that's OK if these methods are only really used in the serialization test.  But it might be worth adding a comment to point out the potential inefficiency.

Reply:
I copied this pattern from `PipelineConfiguration` where config is also stored as `BytesReference`. It is necessary as otherwise equality/hash don't work correctly. I'll add a comment.